### DISCUSSION TITLE
HW0 git clone 身份驗證錯誤解決方案

### DISCUSSION THREAD

**Discussion by lx3211** [1101-GS4719-C/Python-Community-GS4719-C#484] [General]
- 在git clone的時候要求輸入username和password但出現身份驗證錯誤
    - 原因：在 2020 年 7 月的時候 GitHub 就開始聲明將棄用 password authentication 機制，改為採用 token-based authentication 機制。
    - 解決方法：使用 Personal Access Token (PAT)可以與 GitHub 連線
- 建立 PAT
    - 登入 GitHub 後點擊右上角自己的頭像，進入 Settings 中的 `Developer settings`
    - `Personal access tokens → Generate new token`到建立 PAT 的頁面
        - Note：可以視為 PAT 的名稱，目的性越清楚越好，避免一段時間回來看的時候不知其用途。
        -    Expiration：指定 PAT 有效期限，依自己需求設定時間長度。以 Token 來說，一個月到三個月是可接受的範圍，不建議設定為永久。
        - Select scopes：指定 PAT 擁有的權限，每個權限說明可以參考 [Scopes for OAuth Apps](https://docs.github.com/en/developers/apps/building-oauth-apps/scopes-for-oauth-apps)。
    - 填寫好之後點擊 `Generate token` 按鈕產生 PAT
    - 產生的 PAT 會直接顯示在頁面上，需要馬上記下來，否則離開頁面後就無法找回，只能重新產生
- 設置 PAT
    - 直接對 GitHub 使用 push/pull/fetch 等指令，觸發輸入帳密
```bash=
$ git remote add origin https://github.com/{username}/{repository}
$ git push -u origin master
Username for 'https://github.com' : # 輸入你的 GitHub username
Password for 'https://github.com' : # 輸入剛建立好的 PAT，而不是 GitHub 的密碼
```
- 讓 Git 幫忙記住帳號密碼資訊
`$ git config --global credential.helper cache`

**Comment by JSHT** [comment]
這邊如果輸入帳號密碼的方式不行，也可以參考 [StackOverFlow 的方法](https://stackoverflow.com/questions/68775869/support-for-password-authentication-was-removed-please-use-a-personal-access-to)，直接到認證管理員進行編輯。
